sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/ResolveDDLCommandStringTypes.scala

```diff
@@ -28,6 +29,10 @@ import org.apache.spark.sql.types.{DataType, StringType}
  * collation from the corresponding object (table/view -> schema -> catalog).
  */
 object ResolveDDLCommandStringTypes extends Rule[LogicalPlan] {
+  // Tag to mark expressions that have been cast to a new type so that we can
+  // avoid infinite recursion when resolving the same expression multiple times.
+  private val CAST_ADDED_TAG = new TreeNodeTag[Unit]("defaultStringExpressionCastAdded")
```

Comment:
The tag is gone when the nodes are copied IIRC so I don't think this can be used for controlling a behaviour